# DSCI 100 Project: Final Report

**Authors**: 

**Predicting Usage of a Video Game Research Server**
A research group in the Department of Computer Science at UBC (PLAI) has a goal to aid in advancing the limits of artificial intelligence. They host a Minecraft server that records players' actions as they navigate through the world. Since running the project is so complex, they want to recruit players who will contribute large amounts of gameplay data, as well as make sure they have the resources to handle the amount of players they attract. In an attempt to optimize their players' contributions, they want to understand what demographics of players will have the greatest gameplay time. This project answers their question of how player demographics relate to total gameplay time. 

## The Question
**Question 2**: Which kinds of players are most likely to contribute a large amount of data, and how can we identify them for targeted recruitment?

**Predictive Question** Can we predict the total `played_hours` from the `players.csv` dataset based on the two predictors, `experience` and `subscribe`.

**Response variable**:

`played_hours` - total play time and data contribution

**Explanatory variables**:

`experience` - player experience level

`subscribe` -  subscription status of player

The dataset allows us to examine whether `experience` or `subscribe` relates to total playtime. If certain experience levels or subscribtion status tend to play more, those groups may contribute more gameplay data. This helps identify the player demographics most likely to provide substantial data for the research team.


## # Import libraries
import pandas as pd
import altair as alt

# import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_errorData Description: The Players
The dataset used in this analysis is a player demographics and engagement dataset that contains 196 observations (rows) and 9 variables (columns). The dataset records information about individual players, their gaming experience level, subscription status, demographic details, and the amount of time they have spent playing.

- Number of observations: 196 (each row represents one unique player)
- Number of variables: 9
- Observational unit: Player-level data
- Purpose: Identify which demographic characteristics correspond to higher gameplay time (and therefore more data contributed)
  
**Variables**
- `played_hours` (numeric) - Total hours each player spent on the server
- `experience` (categorical) -  Self-reported experience level in Minecraft
- `gender` (categorical) - Player’s gender
- `subscribe` (Boolean) - Subscription status of player
- `hashedEmail` (String) - The players email, hashed for privacy
- `name` (String) - Name of player
- `age` (Numeric) - Age of player
- `individualID` (String) - Unique ID of player
- `organizationName` (String) - Organization name

**Issues to consider**
- Missing identifiers (`individualId`, `organizationName`).
- `hashedEmail` is anonymous for privacy.
- `played_hours` may include idle time, which could overestimate actual engagement.

Source of the original dataset:
https://drive.google.com/file/d/1Mw9vW0hjTJwRWx0bDXiSpYsO3gKogaPz/edit


In [2]:
# Import libraries
import pandas as pd
import altair as alt

# import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

# Read the datasets
url= "https://raw.githubusercontent.com/sydlpeters/dsci-group-2025w1-group-101-1/refs/heads/main/data/players.csv"
# Load the dataset 
players = pd.read_csv(url)
players

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [3]:
# Drop columns that are entirely missing(NaN)
players_tidy = players.drop(columns=["individualId", "organizationName", "name", "gender", "hashedEmail", "age"])
# Preview tidy players_tidy
players_tidy.head()

# Change variables to numeric here

# Split here

experience  subscribe  played_hours
0        Pro       True          30.3
1    Veteran       True           3.8
2    Veteran      False           0.0
3    Amateur       True           0.7
4    Regular       True           0.1

In [4]:
players_tidy.describe()

played_hours
count    196.000000
mean       5.845918
std       28.357343
min        0.000000
25%        0.000000
50%        0.100000
75%        0.600000
max      223.100000

**Summary Statistics**

Most players recorded very little playtime, with three-quarters spending under an hour in total (Q3 = 0.6 hours). Only a small number logged substantially more time, which raises the overall average (mean = 5.85 hours; median = 0.1 hours) and affects the dataset (STD = 28.36 hours). **Most players played very little (close to 0 hours), but a few played way more, making the dataset highly uneven.**

In [5]:
# Experienced comapred with Average played hours played

# Summary statistics and graphs should only be done on training data

plot1= alt.Chart(players, title="Fig 1. Total Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("mean(played_hours):Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1

alt.Chart(...)

In [6]:
# Experienced comapred with Average played hours played
plot1= alt.Chart(players, title="Fig. 2 Average Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("played_hours:Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1

alt.Chart(...)

In [7]:
# Subscription status comapred with Average hours played
plot2= alt.Chart(players, title="Fig. 3 Average Hours by Subscribed/Unsubscribed").mark_bar().encode(
    x=alt.X("subscribe:N").title("Subscription Status"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("subscribe:N").title("Subscription Status")
)
plot2

alt.Chart(...)

In [8]:
plot3= alt.Chart(players, title="Fig. 4 Average Hours Played by Gender").mark_bar().encode(
    x=alt.X("gender:N").title("Gender"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("gender:N", legend=None).title("Gender")
)
plot3

alt.Chart(...)

In [9]:
plot4= alt.Chart(players, title="Fig. 5 Average Hours Played by each Age").mark_circle(opacity=0.8).encode(
    x=alt.X("age:N", scale=alt.Scale(domain=[i for i in range(2, 102, 2)])).title("Age"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("age:N", legend=None).title("Age")
)
plot4

alt.Chart(...)

In [10]:
experience_map = {
    "Beginner" : 0,
    "Regular": 1,
    "Amateur": 2,
    "Veteran": 3,
    "Pro": 4
}
# Put above
players["experience"] = players["experience"].map(experience_map)
players

experience  subscribe                                        hashedEmail  \
0             4       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1             3       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2             3      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3             2       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4             1       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..          ...        ...                                                ...   
191           2       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192           3      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193           2      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194           2      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195           4       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [11]:
subscribe_map = {
    False : 0,
    True : 1,
}

players_tidy["subscribe"] = players_tidy["subscribe"].map(subscribe_map)
players_tidy

experience  subscribe  played_hours
0          Pro          1          30.3
1      Veteran          1           3.8
2      Veteran          0           0.0
3      Amateur          1           0.7
4      Regular          1           0.1
..         ...        ...           ...
191    Amateur          1           0.0
192    Veteran          0           0.3
193    Amateur          0           0.0
194    Amateur          0           2.3
195        Pro          1           0.2

[196 rows x 3 columns]

In [12]:
# split data!
players_training, players_testing = train_test_split(
    players, test_size=0.2, random_state=1234
)

# set target and predictors
players_x_train = players_training[["experience"]]
players_y_train = players_training["played_hours"]

players_x_test = players_testing[["experience"]]
players_y_test = players_testing["played_hours"]

players_training

experience  subscribe                                        hashedEmail  \
114           0       True  ae8d8a9dcf80b38466f89466201e6d594eb94d6994eda1...   
187           2       True  e3f0ad9aadd27f3d1d9197e58546d045018daa76767503...   
39            2       True  6ef84a204d64edc3b3779127b13f298d0c70f96568486f...   
25            1       True  5baba1651a0b92788bc0d6dcdf00be64af1cf9f0015bbe...   
131           2      False  11bf6125c4264b3a8f3bffa57b33bd598e2ea1ecd6331a...   
..          ...        ...                                                ...   
152           0       True  aea049eaa7cb10db386a62990220d205ceb2a4c473cae3...   
116           3      False  58893f3187db90bf0690c88f06e6aa5f8ea9ff9691ca8e...   
53            2       True  fc0224c81384770e93ca717f32713960144bf0b52ff676...   
38            3       True  d782933acd14c834e53dea816005a3583cb87710f7347a...   
47            1      False  453774822d52d6891593948224570a750c4334616ff8aa...   

     played_hours     name             gender  age  individualId  \
114           1.0     Ella             Female   17           NaN   
187           0.0   Jasper               Male   17           NaN   
39            0.0   Vivian               Male   17           NaN   
25            0.6  Kendall             Female   28           NaN   
131           0.0   Olivia             Female   23           NaN   
..            ...      ...                ...  ...           ...   
152           0.2    Aarav  Prefer not to say   17           NaN   
116           0.0     Noah  Prefer not to say   20           NaN   
53            0.2    Gemna               Male   27           NaN   
38            0.0   Ishaan               Male   17           NaN   
47            0.0   Edmund  Prefer not to say   23           NaN   

     organizationName  
114               NaN  
187               NaN  
39                NaN  
25                NaN  
131               NaN  
..                ...  
152               NaN  
116               NaN  
53                NaN  
38                NaN  
47                NaN  

[156 rows x 9 columns]

In [14]:
# preprocess the data, make the pipeline
players_preprocessor = make_column_transformer(
    (
        StandardScaler(),
        [
            "experience",
            ],
    )
)
players_pipeline = make_pipeline(players_preprocessor, KNeighborsRegressor())

players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        players_x_train,
        players_y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)
players_cv

fit_time  score_time  test_score  train_score
0  0.007018    0.001995   -9.099486   -21.719026
1  0.002997    0.000997   -5.920221   -22.051516
2  0.002001    0.001000  -28.129594   -16.957817
3  0.001998    0.001022   -1.713133   -22.352437
4  0.001001    0.000999  -32.491334   -15.254916

In [ ]:
# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)

# fit the GridSearchCV object
players_result = pd.DataFrame(players_tuned.fit(players_x_train, players_y_train).cv_results_)

players_result

/opt/conda/lib/python3.11/site-packages/numpy/ma/core.py:2846: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003402      0.000255         0.002116        0.000136   
1        0.004764      0.002951         0.002155        0.000431   
2        0.019322      0.032565         0.001978        0.000134   
3        0.003059      0.000040         0.001914        0.000008   
4        0.003074      0.000072         0.001927        0.000023   
5        0.003063      0.000048         0.001899        0.000005   
6        0.003034      0.000026         0.001914        0.000021   
7        0.003272      0.000469         0.001898        0.000009   
8        0.003083      0.000118         0.001906        0.000038   
9        0.003062      0.000036         0.001906        0.000020   
10       0.002983      0.000010         0.001888        0.000018   
11       0.003015      0.000055         0.001899        0.000025   
12       0.003130      0.000177         0.002327        0.000830   
13       0.003376      0.000581         0.002047        0.000175   
14       0.003186      0.000264         0.001978        0.000115   
15       0.003039      0.000059         0.001891        0.000028   
16       0.003007      0.000064         0.001878        0.000006   
17       0.002976      0.000018         0.001875        0.000010   
18       0.002983      0.000061         0.001864        0.000007   
19       0.002989      0.000078         0.001870        0.000013   
20       0.002960      0.000027         0.001886        0.000011   
21       0.003157      0.000357         0.001884        0.000004   
22       0.002998      0.000063         0.001880        0.000005   
23       0.002989      0.000060         0.001889        0.000003   
24       0.002971      0.000021         0.001891        0.000030   
25       0.002973      0.000059         0.001874        0.000010   
26       0.002969      0.000060         0.001869        0.000008   
27       0.002956      0.000018         0.001893        0.000017   
28       0.002984      0.000077         0.001876        0.000005   
29       0.002964      0.000051         0.001880        0.000008   
30       0.002992      0.000061         0.001901        0.000027   
31       0.002938      0.000004         0.001883        0.000002   
32       0.002974      0.000055         0.001886        0.000006   
33       0.003005      0.000062         0.001893        0.000007   
34       0.002957      0.000036         0.001922        0.000060   
35       0.003151      0.000353         0.001889        0.000009   
36       0.002996      0.000103         0.001887        0.000011   
37       0.003123      0.000311         0.002086        0.000389   
38       0.002961      0.000034         0.001902        0.000013   
39       0.003017      0.000074         0.001896        0.000005   
40       0.003135      0.000386         0.001914        0.000026   
41       0.002953      0.000014         0.001916        0.000025   
42       0.002973      0.000055         0.001924        0.000027   
43       0.003003      0.000078         0.001913        0.000011   
44       0.002952      0.000027         0.001922        0.000026   
45       0.004269      0.001321         0.002297        0.000404   
46       0.003055      0.000102         0.001949        0.000013   
47       0.003438      0.000230         0.002458        0.000142   
48       0.004011      0.001267         0.002165        0.000190   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [ ]:
# Add Visualization to compare the ks

In [ ]:
# Retrieve the CV scores


# get the best parameter values
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 39}

In [ ]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(16.028531937252072)

In [16]:
knn = KNeighborsRegressor(n_neighbors = 39)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(players_x_train, players_y_train)

players_predictions = players_testing.assign(prediction = player_pipe.predict(players_x_test))
players_predictions

experience  subscribe                                        hashedEmail  \
101           2       True  25879aecc205544bc6505f9faf768356e0a3b712605730...   
51            1       True  b622593d2ef8b337dc554acb307d04a88114f2bf453b18...   
146           4       True  5669c0f4b50dbfe3e2f851e4bb89fa43c55cd1f71ba362...   
153           0       True  0ce7bfa910d47fc91f21a7b3acd8f33bde6db57912ce02...   
106           1      False  63139b524e44d156daf265875fb8078bb2b24d6b7520a3...   
59            3       True  ca20f724571080b997e0efa874b9611e9f280c1af5f68f...   
161           3      False  f174555ae3ff613d5d0b98f83f26f689619d92ef272f28...   
167           0      False  42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898a...   
193           2      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
88            0       True  a1094a440899e69e804f888f44d1154ec0b7675d05d977...   
29            3      False  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b1...   
130           2       True  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...   
66            3      False  70db3a16ef58cf8bf6d8658c426ba4a5fcd7523a0abdd4...   
74            1       True  bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...   
48            3       True  b3510c708bd50bf9f75e6e02bb6fe14edb705e0ea671ee...   
100           2       True  26362473ca8fab46358dbab9bcbcf7114bbcbadef90e6b...   
123           0      False  e74c60a92c0100e7240be56d66969db85856152b048c63...   
31            3       True  f1b432523542f90c61176a555ccb2144468d76c91a32d7...   
95            1      False  dea781ecb85a9e85c8d1fa62b0c2a98c77a51f5933c104...   
137           3       True  d711ff0c9d03664676d1ce80ca7ac5e57f7ebbc0d04bd4...   
6             1       True  8e594b8953193b26f498db95a508b03c6fe1c24bb5251d...   
54            0      False  5e5c25a773be7a62638a163d773534e575a5ad57821047...   
97            3      False  d3ca24e4d7fe8ffe2a821ebd3b841252950ca53e4d659a...   
5             2       True  f58aad5996a435f16b0284a3b267f973f9af99e7a89bee...   
110           2       True  827f9a82b47cdf21098bc7bb1d4241550dc6146ba9ae8b...   
91            0       True  eeb961422a3f9841018d4eb729ddc40ef103cc1205577d...   
134           0       True  fef4e1bed8c3f6dcd7bcd39ab21bd402386155b2ff8c8e...   
180           2      False  93efabc9f7de58f44c457bac715cf8f9f44cd88cc574fc...   
7             2      False  1d2371d8a35c8831034b25bda8764539ab7db0f6393869...   
168           3       True  9ba67e127e4e70d909f6237644bb4674b17bc6ae5c9e73...   
94            0       True  e3175c379bbb83c8786d2dd719cf844575aca24d2c81d6...   
9             3       True  bbe2d83de678f519c4b3daa7265e683b4fe2d814077f90...   
58            1       True  f2826fb8dbce4d450348f99cb27ade184b713998d96797...   
92            4       True  fcab03c6d3079521e7f9665caed0f31fe3dae6b5ccb86e...   
111           1       True  336adf9b9070abb44eaec22a48b04d8cda6cf0c98a46ee...   
43            3       True  126ec350278422ba60ad575d230c19522fd568094b7c0f...   
77            1       True  e21a324ccf5c873bafe82e47d5137b36aa312ee4803eeb...   
142           0       True  6ca8a90e4f5c61bf36baf556d446873c4e894e7a62f919...   
119           0       True  d7e7a5ff16f94d1e793490b521671f7ca56d77820df9b9...   
40            1       True  0d4d71be33e2bc7266ee4983002bd930f69d304288a866...   

     played_hours       name             gender  age  individualId  \
101           0.0        Eli             Female   17           NaN   
51          218.1       Akio         Non-binary   20           NaN   
146           0.0      Padma         Non-binary   25           NaN   
153           0.1     Osiris               Male   17           NaN   
106           0.0      Elias               Male   23           NaN   
59            0.2     Edward               Male   38           NaN   
161           0.0     Finley         Non-binary   17           NaN   
167           0.3     Ariana             Female   17           NaN   
193           0.0      Dylan  Prefer not to say   17 

In [ ]:

players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=39"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction"
))
players_plot

# RMSE on the training add.

alt.LayerChart(...)

In [ ]:
players_predictions = players_predictions.rename(columns={"prediction":"prediction_80"})

players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        players_x_train,
        players_y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)

# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)



players_result = pd.DataFrame(players_tuned.fit(players[["experience"]], players[["played_hours"]]).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003600      0.000201         0.003483        0.002256   
1        0.003519      0.000046         0.002358        0.000055   
2        0.003441      0.000018         0.002311        0.000016   
3        0.003602      0.000213         0.003073        0.001017   
4        0.003428      0.000038         0.002312        0.000024   
5        0.003444      0.000061         0.002298        0.000013   
6        0.004929      0.003024         0.002314        0.000030   
7        0.003450      0.000059         0.002294        0.000019   
8        0.003445      0.000060         0.002300        0.000016   
9        0.003415      0.000046         0.002350        0.000107   
10       0.003484      0.000107         0.002305        0.000010   
11       0.003407      0.000016         0.002306        0.000010   
12       0.003430      0.000071         0.002309        0.000018   
13       0.003409      0.000019         0.002301        0.000009   
14       0.003603      0.000332         0.002312        0.000032   
15       0.003669      0.000406         0.002380        0.000064   
16       0.003491      0.000089         0.002341        0.000013   
17       0.003458      0.000051         0.002334        0.000016   
18       0.003652      0.000364         0.002324        0.000007   
19       0.003442      0.000052         0.002316        0.000011   
20       0.003525      0.000155         0.002339        0.000028   
21       0.003493      0.000112         0.002330        0.000015   
22       0.003419      0.000018         0.002324        0.000011   
23       0.003441      0.000066         0.002350        0.000065   
24       0.003408      0.000017         0.002333        0.000007   
25       0.003421      0.000068         0.002320        0.000014   
26       0.003429      0.000056         0.002327        0.000009   
27       0.003401      0.000021         0.002325        0.000009   
28       0.003449      0.000075         0.002318        0.000014   
29       0.003381      0.000014         0.002333        0.000004   
30       0.003412      0.000045         0.002322        0.000008   
31       0.003376      0.000012         0.002331        0.000019   
32       0.003421      0.000076         0.002342        0.000033   
33       0.003395      0.000013         0.002334        0.000016   
34       0.003432      0.000047         0.002344        0.000031   
35       0.003443      0.000059         0.002400        0.000105   
36       0.003385      0.000012         0.002335        0.000013   
37       0.003451      0.000070         0.002342        0.000011   
38       0.003412      0.000047         0.002347        0.000014   
39       0.003431      0.000044         0.002340        0.000009   
40       0.003390      0.000011         0.002362        0.000012   
41       0.003448      0.000080         0.002374        0.000031   
42       0.003399      0.000014         0.002353        0.000010   
43       0.003440      0.000067         0.002364        0.000024   
44       0.003486      0.000046         0.002394        0.000016   
45       0.003403      0.000031         0.002354        0.000013   
46       0.003418      0.000057         0.002352        0.000016   
47       0.003423      0.000046         0.002391        0.000039   
48       0.003405      0.000065         0.002359        0.000012   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [ ]:
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 40}

In [ ]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(24.74677992806886)

In [ ]:
knn = KNeighborsRegressor(n_neighbors = 12)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(players[["experience"]], players[["played_hours"]])

print(players_predictions)

players_predictions = players.assign(prediction_all = player_pipe.predict(players[["experience"]]))
players_predictions

     experience  subscribe                                        hashedEmail  \
101           2          1  25879aecc205544bc6505f9faf768356e0a3b712605730...   
51            1          1  b622593d2ef8b337dc554acb307d04a88114f2bf453b18...   
146           4          1  5669c0f4b50dbfe3e2f851e4bb89fa43c55cd1f71ba362...   
153           0          1  0ce7bfa910d47fc91f21a7b3acd8f33bde6db57912ce02...   
106           1          0  63139b524e44d156daf265875fb8078bb2b24d6b7520a3...   
59            3          1  ca20f724571080b997e0efa874b9611e9f280c1af5f68f...   
161           3          0  f174555ae3ff613d5d0b98f83f26f689619d92ef272f28...   
167           0          0  42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898a...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
88            0          1  a1094a440899e69e804f888f44d1154ec0b7675d05d977...   
29            3          0  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b1...   
130           2          1  

experience  subscribe                                        hashedEmail  \
0             4          1  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1             3          1  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2             3          0  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3             2          1  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4             1          1  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..          ...        ...                                                ...   
191           2          1  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192           3          0  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194           2          0  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195           4          1  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  prediction_all  
0                 NaN        0.491667  
1                 NaN        0.625000  
2                 NaN        0.625000  
3                 NaN        4.475000  
4                 NaN       19.208333  
..                ...             ...  
191               NaN        4.475000  
192               NaN        0.625000  
193               NaN        4.475000  
194               NaN        4.475000  
195               NaN        0.491667  

[196 rows x 10 columns]

In [ ]:

players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=12"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction_all"
))
players_plot

alt.LayerChart(...)

In [ ]:
# Train only on the training dataset (includes cross validation), predict on the whole dataset